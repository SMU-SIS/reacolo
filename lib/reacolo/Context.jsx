import React from 'react';

const Context = () => {
  throw new Error('Context components cannot be rendered directly. They must be contained either ' +
    'in a Filter component or a parent Context component.');
};

Context.propTypes = {
  default: React.PropTypes.bool,
  children: React.PropTypes.node
};

Context.defaultProps = {
  default: false
};

export default Context;
