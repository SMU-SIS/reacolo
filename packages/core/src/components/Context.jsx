import React from 'react';
// import omit from 'object.omit';
import PropTypes from 'prop-types';
import matchContext from '../filtering/match-context.js';
import createMatcher from '../filtering/create-matcher.js';
import connect from '../connect/connect.jsx';

/**
 * Context component. Renders only if its value matches its match properties.
 * @param {object} props The component's properties.
 * @return {Element} A react element.
 */
export const Context = props => {
  const matcher = createMatcher(props);
  const shouldRender = matchContext(props.contextValue, matcher);
  if (!shouldRender) return null;
  if (props.render) return props.render();
  if (props.component) return <props.component />;
  if (props.children) return props.children;
  throw new Error(
    'No render function nor component properties have been provided to <Context>',
  );
};

Context.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  contextValue: PropTypes.object.isRequired,
  // eslint-disable-next-line react/no-unused-prop-types
  default: PropTypes.bool,
  render: PropTypes.func,
  component: PropTypes.func,
  children: PropTypes.element,
};

Context.defaultProps = {
  default: false,
  render: undefined,
  component: undefined,
  children: undefined,
};

// Default export is a <ConnectedContext>, a <Context> that fetches its value
// automatically from the React context.
export default connect(undefined, undefined, context => ({
  contextValue: context,
}))(Context);
