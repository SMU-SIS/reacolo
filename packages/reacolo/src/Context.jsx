import bindContext from './bind-context';

export default bindContext();
