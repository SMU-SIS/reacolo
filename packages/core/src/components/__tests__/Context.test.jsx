import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import { Context } from '../Context.jsx';
import matchContext from '../../filtering/match-context.js';

jest.mock('../../filtering/match-context.js');

Enzyme.configure({ adapter: new EnzymeAdapter() });

describe('Context', () => {
  beforeEach(() => {
    matchContext.mockReset();
    matchContext.mockImplementation(() => false);
  });

  it('properly calls `matchContext` with the context and the properties to match', () => {
    shallow(
      <Context
        prop1="ignore me"
        matchTest="test value"
        match={{ a: 'a value', b: { bProp: 'value' } }}
        prop2="ignore me again"
        matchOtherTest={{ other: 'test' }}
        render={() => <div />}
        value={{ contextProp: 'val' }}
      />,
    );
    expect(matchContext.mock.calls).toEqual([
      [
        { contextProp: 'val' },
        {
          test: 'test value',
          a: 'a value',
          b: { bProp: 'value' },
          otherTest: { other: 'test' },
        },
      ],
    ]);
  });

  it('does not render if `matchContext returns false`', () => {
    const wrapper = shallow(
      <Context
        render={() => <div>test</div>}
        matchProp="val"
        value={{ contextProp: 'val' }}
      />,
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('renders using its component property if `matchContext` returns true and it has no render function', () => {
    const Child = () => <div>test</div>;
    matchContext.mockImplementation(() => true);
    const wrapper = shallow(
      <Context
        component={Child}
        matchProp="val"
        value={{ contextProp: 'val' }}
      />,
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('renders using its render function if it is provided and if `matchContext` returns true', () => {
    matchContext.mockImplementation(() => true);
    const wrapper = shallow(
      <Context
        render={() => <div>test</div>}
        matchProp="val"
        value={{ contextProp: 'val' }}
      />,
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('throws if there is neither a render function nor a component property regardless of `matchContext` results', () => {
    matchContext.mockImplementation(() => true);
    expect(() => {
      shallow(<Context matchProp="val" value={{ contextProp: 'val' }} />);
    }).toThrow();
  });
});
