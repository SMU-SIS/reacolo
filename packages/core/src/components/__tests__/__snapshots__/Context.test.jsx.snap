// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context does not render if \`matchContext returns false\` 1`] = `""`;

exports[`Context renders in accordance with the computedMatchResult prop if it is provided 1`] = `""`;

exports[`Context renders in accordance with the computedMatchResult prop if it is provided 2`] = `<div />`;

exports[`Context renders its children if there is any, \`matchContext\` returns true and there is no render function nor component property 1`] = `
<div>
  render that
</div>
`;

exports[`Context renders using its component property if \`matchContext\` returns true and it has no render function 1`] = `<component />`;

exports[`Context renders using its render function if it is provided and if \`matchContext\` returns true 1`] = `
<div>
  render that
</div>
`;
